lightly/models/spark.py

```diff
@@ -0,0 +1,280 @@
+        mask_ratio:
+            Ratio of the image to mask. Default: 0.6
+        densify_norm:
+            Type of normalization to use for densification. Default: 'bn'
+        sbn:
+            Whether to use SyncBatchNorm. Default: False
+    """
+    def __init__(
+            self, sparse_encoder: SparseResnet, dense_decoder: LightDecoder,
+            mask_ratio: float=0.6, densify_norm: Literal['batch_norm', 'layer_norm', 'identity']='bn', sbn: bool=False,
+    ):
+        super().__init__()
+        input_size, downsample_ratio = sparse_encoder.input_size, sparse_encoder.downsample_ratio
+        self.downsample_ratio = downsample_ratio
+        self.fmap_h, self.fmap_w = input_size // downsample_ratio, input_size // downsample_ratio
+        self.mask_ratio = mask_ratio
+        self.len_keep = round(self.fmap_h * self.fmap_w * (1 - mask_ratio))
+        
+        self.sparse_encoder = sparse_encoder
+        self.dense_decoder = dense_decoder
+        
+        self.sbn = sbn
+        self.hierarchy = len(sparse_encoder.enc_feat_map_chs)
+        self.densify_norm_str = densify_norm.lower()
+        self.densify_norms = nn.ModuleList()
+        self.densify_projs = nn.ModuleList()
+        self.mask_tokens = nn.ParameterList()
+        
+        e_widths, d_width = self.sparse_encoder.enc_feat_map_chs, self.dense_decoder.width
+        e_widths: List[int]
+        for i in range(self.hierarchy):
+            e_width = e_widths.pop()
+            p = nn.Parameter(torch.zeros(1, e_width, 1, 1))
+            torch.nn.init.trunc_normal_(p, mean=0, std=.02, a=-.02, b=.02)
+            self.mask_tokens.append(p)
+            
+            
+            if self.densify_norm_str == 'batch_norm':
+                densify_norm = (SparseSyncBatchNorm2d if self.sbn else SparseBatchNorm2d)(e_width)
+            elif self.densify_norm_str == 'layer_norm':
+                densify_norm = SparseConvNeXtLayerNorm(e_width, data_format='channels_first', sparse=True)
+            else:
+                densify_norm = nn.Identity()
+            self.densify_norms.append(densify_norm)
+            
+            if i == 0 and e_width == d_width:
+                densify_proj = nn.Identity()   
+                print(f'[SparK.__init__, densify {i+1}/{self.hierarchy}]: use nn.Identity() as densify_proj')
+            else:
+                kernel_size = 1 if i <= 0 else 3
+                densify_proj = nn.Conv2d(e_width, d_width, kernel_size=kernel_size, stride=1, padding=kernel_size // 2, bias=True)
+                print(f'[SparK.__init__, densify {i+1}/{self.hierarchy}]: densify_proj(ksz={kernel_size}, #para={sum(x.numel() for x in densify_proj.parameters()) / 1e6:.2f}M)')
+            self.densify_projs.append(densify_proj)
+            d_width //= 2
+        
+        print(f'[SparK.__init__] dims of mask_tokens={tuple(p.numel() for p in self.mask_tokens)}')
+        
+    
+    def mask(self, B: int, device, generator=None):
```

Comment:
I am not sure how the masking works exactly but we already have a bunch of masking functions here https://github.com/lightly-ai/lightly/blob/356ae56fd8f6d6bbc548d4d189d9444f780a62ca/lightly/models/utils.py#L571-L838

If none of the existing functions are compatible then please add a new function there.